src/esp/gfx/Shader.h

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <Magnum/GL/AbstractShaderProgram.h>
+
+#include "esp/core/esp.h"
+
+namespace esp {
+namespace gfx {
+
+class Drawable;
+class DrawableGroup;
+class RenderCamera;
+
+// TODO: string type to allow for dynamically added shader types
+enum class ShaderType {
+  /**
+   * Shader program for instance mesh data. See @ref gfx::PrimitiveIDShader,
+   * @ref GenericInstanceMeshData, @ref Mp3dInstanceMeshData, @ref
+   * AssetType::INSTANCE_MESH, @ref loadInstanceMeshData.
+   */
+  INSTANCE_MESH_SHADER = 0,
+
+  /**
+   * Shader program for PTex mesh data. See @ref gfx::PTexMeshShader, @ref
+   * gfx::PTexMeshDrawable, @ref loadPTexMeshData, @ref PTexMeshData.
+   */
+  PTEX_MESH_SHADER = 1,
+
+  /**
+   * Shader program for flat shading with uniform color. Used to render object
+   * identifier or semantic types (e.g. chair, table, etc...). Also the
+   * default shader for assets with unidentified rendering parameters. See
+   * @ref Magnum::Shaders::Flat3D.
+   */
+  COLORED_SHADER = 2,
+
+  /**
+   * Shader program for vertex color shading. Used to render meshes with
+   * per-vertex colors defined.
+   */
+  VERTEX_COLORED_SHADER = 3,
+
+  /**
+   * Shader program for meshes with textured defined.
+   */
+  TEXTURED_SHADER = 4,
+
+  COLORED_SHADER_PHONG = 5,
+  VERTEX_COLORED_SHADER_PHONG = 6,
+  TEXTURED_SHADER_PHONG = 7
+};
+
+struct ShaderConfiguration {
+  ShaderType type = ShaderType::TEXTURED_SHADER;
+  // JSON options;
```

Comment:
there won't be derived configs, instead the factories will all take in JSON options, which will be converted to specific config types based in the specific factory that gets called, before the config is used to construct the ShaderProgram. This is quite flexible since it doesn't require complex inheritance hierarchies, and makes it possible for new configurations to be specified in python/javascript. This is the same approach being taken in the Sensor API redesign.